### PR TITLE
Support PointData writing.

### PR DESCRIPTION
**Pull request by feuerste**
With this PR we have the possibility to also write `PointData`, which is more generic than only writing `Point`s of position, color, and intensity.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -71,7 +72,11 @@ impl AttributeData {
...
     pub position: Vec<Vector3<f64>>,
-    pub attributes: FnvHashMap<String, AttributeData>,
+    // BTreeMap for deterministic iteration order.
+    pub attributes: BTreeMap<String, AttributeData>,
 }
 
 pub use point_viewer_proto_rust::proto;
+
+/// size for batch
+pub const NUM_POINTS_PER_BATCH: usize = 500_000;
```
Comment (catevita) [inline src/lib.rs]: Nit: as this refers to the size of PointData and not anymore to the Batch Iterator, could the name reflect more this change

File: src/octree/octree_iterator.rs [hunk 2]
```diff
@@ -36,7 +36,7 @@ impl NodeIterator<CubeNodeReader> {
         }
 
         if let Ok(mut data_map) = octree_data_provider.data(id, &["intensity"]) {
-            match data_map.remove("position") {
+            match data_map.remove("intensity") {
```
Comment (nnmm) [inline src/octree/octree_iterator.rs]: Is that a fix for something that slipped in by accident in a previous PR?

Comment (feuerste) [reply src/octree/octree_iterator.rs]: Yes, sorry, I wanted to mention it... I can also make a mini PR for this, if you prefer...

Comment (nnmm) [reply src/octree/octree_iterator.rs]: I don't think it's a problem, I was just curious :)

Comment (feuerste) [reply src/octree/octree_iterator.rs]: Ok

File: src/lib.rs (hunk 1 above)
Comment (feuerste) [reply src/lib.rs]: @nnmm @catevita Or we even rename `PointData` to `PointBatch`? What do you think?

Comment (nnmm) [reply src/lib.rs]: +1

File: src/read_write/cube.rs [hunk 3]
```diff
@@ -163,71 +164,104 @@ pub struct CubeNodeWriter {
...
+                    vec3_encode::<f64>(position, min, edge_length)
+                        .write_le(&mut self.xyz_writer)?;
+                }
             }
         }
 
-        p.color.write_le(&mut self.attribute_writers[0]);
-        if let Some(i) = p.intensity {
-            if self.attribute_writers.len() < 2 {
+        if self.point_count == 0 {
```
Comment (nnmm) [inline src/read_write/cube.rs]: Does PointData always contain > 0 points?

**Comment by nnmm** [comment]
Basic question: The primary motivation for this approach over something like https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=f373b5ffbdc37377a7a5caad663ad9cd is performance, right?

File: src/read_write/cube.rs (hunk 3 above)
Comment (feuerste) [reply src/read_write/cube.rs]: Theoretically it can contain anything from 0 to NUM_POINTS_PER_BATCH...

**Comment by feuerste** [comment]
> Basic question: The primary motivation for this approach over something like https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=f373b5ffbdc37377a7a5caad663ad9cd is performance, right?

Yes, here we have far less indirections and still have the possibility to either write "old" points or new points batches...

File: src/lib.rs (hunk 1 above)
Comment (feuerste) [reply src/lib.rs]: Done, I renamed it to `PointsBatch`

File: src/read_write/cube.rs (hunk 3 above)
Comment (nnmm) [reply src/read_write/cube.rs]: Maybe this should test `self.attribute_writers.len()` instead?

Comment (feuerste) [reply src/read_write/cube.rs]: Unfortunately this would repeat the check for each iteration, if we don't write any attributes.

Comment (feuerste) [reply src/read_write/cube.rs]: This would be possible, but would be slightly slower, as we needed two check when writing no attributes at all.

Comment (feuerste) [reply src/read_write/cube.rs]: Hm, but since we almost always write attributes, it is nicer, since we can get rid of `point_count`.

**Comment by nnmm** [comment]
That's true, because `write()` does reflection: looping over a str representation of the attributes and calling a 'get' method for each one which returns a general 'attribute' datatype that has to be matched against.

On the other hand, the efficiency in this approach comes from making that reflection less frequent (once per batch), but does that work well with the generation code for octrees? There we usually deal with single points, and you mentioned once that it is not straightforward to translate this code to use batches.

File: src/read_write/cube.rs (hunk 3 above)
Comment (feuerste) [reply src/read_write/cube.rs]: Done, with `self.attribute_writers.is_empty()`

Comment (nnmm) [reply src/read_write/cube.rs]: :+1:

**Comment by nnmm** [comment]
Summary of offline discussion: There is no good way around reflection on single points during node splitting. There we'll need either an `impl NodeWriter<DynamicPoint> for CubeNodeWriter` or some reflection to put a point into another `PointsBatch`.

**Comment by nnmm** [comment]
For now, we'll postpone arbitrary attributes in octree generation and continue to work with `Point`s.

**Comment by nnmm** [comment]
One last high-level comment: What about having a `write()` and `write_batch()` method instead of a type parameter?

**Comment by feuerste** [comment]
> One last high-level comment: What about having a `write()` and `write_batch()` method instead of a type parameter?

This is a matter of taste. I actually had this before, but personally preferred a generic.

**Comment by feuerste** [comment]
@wally-the-cartographer merge

**Comment by feuerste** [comment]
Thanks for the review Nikolai!

**Comment by wally-the-cartographer** [comment]
Merge requested by authorized user feuerste. Merge queue now has a length of 1.

**Pull request merged**